apps/studio/src/server/modules/audit/audit.service.ts

```diff
@@ -0,0 +1,181 @@
+import type {
+  AuditLogEvent,
+  Blob,
+  DB,
+  Resource,
+  ResourcePermission,
+  Transaction,
+  User,
+  VerificationToken,
+} from "../database"
+
+type FullResource = Resource & (Blob | undefined)
+
+interface CreateDelta {
+  before: null
+  after: FullResource
+}
+interface DeleteDelta {
+  before: FullResource
+  after: null
+}
+interface UpdateDelta {
+  before: FullResource
+  after: FullResource
+}
+
+interface ResourceEventLogProps {
+  eventType: Extract<
+    AuditLogEvent,
+    "ResourceCreate" | "ResourceUpdate" | "ResourceDelete" | "ResourceMove"
+  >
+  delta: CreateDelta | DeleteDelta | UpdateDelta
+  by: User
+  ip?: string
+}
+
+// NOTE: Type to force every logger method to have a tx
+type AuditLogger<T> = (tx: Transaction<DB>, props: T) => Promise<void>
+
+export const logResourceEvent: AuditLogger<ResourceEventLogProps> = async (
+  tx,
+  { eventType, delta, by, ip },
+) => {
+  await tx
+    .insertInto("AuditLog")
+    .values({ eventType, delta, userId: by.id, ipAddress: ip })
+    .execute()
+}
+
+interface LoginDelta {
+  before: VerificationToken
+  after: VerificationToken
```

Comment:
suggestion: was thinking perhaps we should also use `User` in the login to make it similar to the logout? This might make it more logical when comparing between the session creation and session destruction.

Separately, I think we might need to expand what the User object has to capture the session information, because it might be possible for a user to create multiple sessions across different devices. Ideally during login, we can give the session an ID, then use it to identify which session is being destroyed later.

Reply:
the `VerificationToken` here consists of `user.email` with the `ip` (see ss). because of this, it's possible for the user to have multiple session over devices **but** this will be captured under teh `ip` section (distinct ip) though there is a short-coming in that the user isn't logged directly; i'll add it in!

<img width="1182" alt="image" src="https://github.com/user-attachments/assets/aa0a832f-115e-46c6-afe6-6c39e8c11528" />